pdearena/modules/threed.py

```diff
@@ -0,0 +1,201 @@
+            ):
+        """Initialize basic 3d FNO ResNet building block
+        Args:
+            in_planes (int): input channels
+            planes (int): output channels
+            stride (int, optional): stride of 2d convolution. Defaults to 1.
+        """
+        super().__init__()
+
+        self.fourier1 = SpectralConv3d(
+            in_planes, planes, modes1=modes1, modes2=modes2, modes3=modes3
+        )
+        self.conv1 = nn.Conv3d(in_planes, planes, kernel_size=1, padding=0, padding_mode="zeros")
+        self.fourier2 = SpectralConv3d(
+            planes, planes, modes1=modes1, modes2=modes2, modes3=modes3
+        )
+        self.conv2 = nn.Conv3d(planes, planes, kernel_size=1, padding=0, padding_mode="zeros")
+
+        # Shortcut connection, batchnorm removed
+        # So far shortcut connections are not helping
+        self.shortcut = nn.Sequential()
+        if stride != 1 or in_planes != self.expansion * planes:
+            self.shortcut = nn.Sequential(
+                nn.Conv2d(in_planes, self.expansion * planes, kernel_size=1)
+            )
+        if activation == "gelu":
+            self.activation = F.gelu
+        elif activation == "relu":
+            self.activation = F.relu
+        else:
+            raise NotImplementedError(f"Activation {activation} not implemented")
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward pass of basic 3d Fourier ResNet building block.
+        Args:
+            x (torch.Tensor): input of shape [batch, in_planes, x, y, z]
+        Returns:
+            torch.Tensor: output of shape [batch, planes, x, y, z]
+        """
+        x1 = self.fourier1(x)
+        x2 = self.conv1(x)
+        out = self.activation(x1 + x2)
+
+        x1 = self.fourier2(out)
+        x2 = self.conv2(out)
+        out = x1 + x2
+        # out += self.shortcut(x)
+        out = self.activation(out)
+        return out
+
+
+class ResNet3D(nn.Module):
+    """3d ResNet model, difference to default ResNet architectures is that
+    spatial resolution and channels (in_planes) stay constant throughout the network."""
+
+    padding = 2  # no periodic
+
+    def __init__(
+        self,
+        block: nn.Module,
+        num_blocks: list,
+        time_history: int,
+        time_future: int,
+        hidden_channels: int = 64,
+        activation: str = "gelu",
+        diffmode: bool = False,
+    ):
+        """Initialize 3d ResNet model
+
+        Args:
+            block (nn.Module): basic 3d ResNet building block
+            num_blocks (list): list of basic building blocks per layer
+            time_history (int): input timesteps
+            time_future (int): prediction timesteps
+            hidden_channels (int): hidden channels in the ResNet blocks
+        """
+        super(ResNet3D, self).__init__()
+
+        self.diffmode = diffmode
+        self.in_planes = hidden_channels
+        self.conv_in1 = nn.Conv3d(
+            time_history * 6,
```

Comment:
Should we call this MaxwellResNet3D as well? As for other problems the input channel could have been other than 6? Sorry for the late review.

Reply:
yes this is a good catch